// @nats-io/nats-core/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleMutex = exports.Perf = void 0;
exports.extend = extend;
exports.render = render;
exports.timeout = timeout;
exports.delay = delay;
exports.deadline = deadline;
exports.deferred = deferred;
exports.debugDeferred = debugDeferred;
exports.shuffle = shuffle;
exports.collect = collect;
exports.jitter = jitter;
exports.backoff = backoff;
exports.nanos = nanos;
exports.millis = millis;
exports.randomToken = randomToken;
/*
 * Copyright 2018-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// deno-lint-ignore-file no-explicit-any
const encoders_1 = require("./encoders");
const errors_1 = require("./errors");
function extend(a, ...b) {
    for (let i = 0; i < b.length; i++) {
        const o = b[i];
        Object.keys(o).forEach(function (k) {
            a[k] = o[k];
        });
    }
    return a;
}
function render(frame) {
    const cr = "␍";
    const lf = "␊";
    return encoders_1.TD.decode(frame)
        .replace(/\n/g, lf)
        .replace(/\r/g, cr);
}
function timeout(ms, asyncTraces = true) {
    // by generating the stack here to help identify what timed out
    const err = asyncTraces ? new errors_1.TimeoutError() : null;
    let methods;
    let timer;
    const p = new Promise((_resolve, reject) => {
        const cancel = () => {
            if (timer) {
                clearTimeout(timer);
            }
        };
        methods = { cancel };
        // @ts-ignore: node is not a number
        timer = setTimeout(() => {
            if (err === null) {
                reject(new errors_1.TimeoutError());
            }
            else {
                reject(err);
            }
        }, ms);
    });
    // noinspection JSUnusedAssignment
    return Object.assign(p, methods);
}
function delay(ms = 0) {
    let methods;
    const p = new Promise((resolve) => {
        const timer = setTimeout(() => {
            resolve();
        }, ms);
        const cancel = () => {
            if (timer) {
                clearTimeout(timer);
                resolve();
            }
        };
        methods = { cancel };
    });
    return Object.assign(p, methods);
}
async function deadline(p, millis = 1000) {
    const d = deferred();
    const timer = setTimeout(() => {
        d.reject(new errors_1.TimeoutError());
    }, millis);
    try {
        return await Promise.race([p, d]);
    }
    finally {
        clearTimeout(timer);
    }
}
/**
 * Returns a Promise that has a resolve/reject methods that can
 * be used to resolve and defer the Deferred.
 */
function deferred() {
    let methods = {};
    const p = new Promise((resolve, reject) => {
        methods = { resolve, reject };
    });
    return Object.assign(p, methods);
}
function debugDeferred() {
    let methods = {};
    const p = new Promise((resolve, reject) => {
        methods = {
            resolve: (v) => {
                console.trace("resolve", v);
                resolve(v);
            },
            reject: (err) => {
                console.trace("reject");
                reject(err);
            },
        };
    });
    return Object.assign(p, methods);
}
function shuffle(a) {
    for (let i = a.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [a[i], a[j]] = [a[j], a[i]];
    }
    return a;
}
async function collect(iter) {
    const buf = [];
    for await (const v of iter) {
        buf.push(v);
    }
    return buf;
}
class Perf {
    timers;
    measures;
    constructor() {
        this.timers = new Map();
        this.measures = new Map();
    }
    mark(key) {
        this.timers.set(key, performance.now());
    }
    measure(key, startKey, endKey) {
        const s = this.timers.get(startKey);
        if (s === undefined) {
            throw new Error(`${startKey} is not defined`);
        }
        const e = this.timers.get(endKey);
        if (e === undefined) {
            throw new Error(`${endKey} is not defined`);
        }
        this.measures.set(key, e - s);
    }
    getEntries() {
        const values = [];
        this.measures.forEach((v, k) => {
            values.push({ name: k, duration: v });
        });
        return values;
    }
}
exports.Perf = Perf;
class SimpleMutex {
    max;
    current;
    waiting;
    /**
     * @param max number of concurrent operations
     */
    constructor(max = 1) {
        this.max = max;
        this.current = 0;
        this.waiting = [];
    }
    /**
     * Returns a promise that resolves when the mutex is acquired
     */
    lock() {
        // increment the count
        this.current++;
        // if we have runners, resolve it
        if (this.current <= this.max) {
            return Promise.resolve();
        }
        // otherwise defer it
        const d = deferred();
        this.waiting.push(d);
        return d;
    }
    /**
     * Release an acquired mutex - must be called
     */
    unlock() {
        // decrement the count
        this.current--;
        // if we have deferred, resolve one
        const d = this.waiting.pop();
        d?.resolve();
    }
}
exports.SimpleMutex = SimpleMutex;
/**
 * Returns a new number between  .5*n and 1.5*n.
 * If the n is 0, returns 0.
 * @param n
 */
function jitter(n) {
    if (n === 0) {
        return 0;
    }
    return Math.floor(n / 2 + Math.random() * n);
}
/**
 * Returns a Backoff with the specified interval policy set.
 * @param policy
 */
function backoff(policy = [0, 250, 250, 500, 500, 3000, 5000]) {
    if (!Array.isArray(policy)) {
        policy = [0, 250, 250, 500, 500, 3000, 5000];
    }
    const max = policy.length - 1;
    return {
        backoff(attempt) {
            return jitter(attempt > max ? policy[max] : policy[attempt]);
        },
    };
}
/**
 * Converts the specified millis into Nanos
 * @param millis
 */
function nanos(millis) {
    return millis * 1000000;
}
/**
 * Convert the specified Nanos into millis
 * @param ns
 */
function millis(ns) {
    return Math.floor(ns / 1000000);
}
const tokenDigits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const tokenDigitCodes = new Uint8Array(62);
for (let i = 0; i < 62; i++)
    tokenDigitCodes[i] = tokenDigits.charCodeAt(i);
// 62^8 = 218,340,105,584,896 — fits safely in Number.MAX_SAFE_INTEGER (2^53-1).
const tokenSpace = 218340105584896;
/**
 * Returns an 8-char base62 token from `Math.random()`. Used as the per-request
 * suffix on a shared `_INBOX.<nuid>.*` mux subscription — mirrors the
 * `replySuffixLen=8` token generated by nats.go in `(nc *Conn).newRespInbox`.
 *
 * Not cryptographically random and does not check for collisions. Collision
 * avoidance relies on (a) the per-connection nuid prefix isolating namespace
 * across connections, and (b) the in-flight token set being small relative
 * to the 62^8 ≈ 2.18e14 space.
 */
function randomToken() {
    let n = Math.floor(Math.random() * tokenSpace);
    let d = n % 62;
    const c0 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c1 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c2 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c3 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c4 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c5 = tokenDigitCodes[d];
    n = (n - d) / 62;
    d = n % 62;
    const c6 = tokenDigitCodes[d];
    n = (n - d) / 62;
    const c7 = tokenDigitCodes[n];
    return String.fromCharCode(c0, c1, c2, c3, c4, c5, c6, c7);
}
